web/pdf_save_service.js

```diff
@@ -0,0 +1,64 @@
+/* Copyright 2016 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PDFSaveServiceFactory } from "./app.js";
+
+let activeSaveService = null;
+
+function PDFSaveService(pdfDocument, pagesCount) {
+  this.pdfDocument = pdfDocument;
+  this.pagesCount = pagesCount;
+}
+
+PDFSaveService.prototype = {
+  saveDocument() {
+    return this.pdfDocument.saveDocument(
+      this.pagesCount,
+      this.pdfDocument.annotationStorage.getAll()
+    );
+  },
+
+  get active() {
+    return this === activeSaveService;
+  },
+
+  destroy() {
+    if (activeSaveService !== this) {
+      // |activeService| cannot be replaced without calling destroy() first,
+      // so if it differs then an external consumer has a stale reference to
+      // us.
+      return;
+    }
+    activeSaveService = null;
+  },
+
+  throwIfInactive() {
+    if (!this.active) {
+      throw new Error("This save request was cancelled or completed.");
+    }
+  },
+};
+
+PDFSaveServiceFactory.instance = {
+  createSaveService(pdfDocument, pagesCount) {
+    if (activeSaveService) {
+      throw new Error("The save service is created and active.");
+    }
+    activeSaveService = new PDFSaveService(pdfDocument, pagesCount);
+    return activeSaveService;
+  },
+};
+
+export { PDFSaveService };
```

Comment:
I don't think this is/should be used elsewhere because the factory gives you an instance, so this should probably be removed.